assignments/assignment02.rb

```diff
@@ -0,0 +1,332 @@
+  
+end
+
+# Your method should generate the following results:
+to_sentence []                       #=> ""
+to_sentence ["john"]                 #=> "john"
+to_sentence ["john", "paul"]         #=> "john and paul"
+to_sentence [1, "paul", 3, "ringo"]  #=> "1, paul, 3 and ringo"
+
+
+# ========================================================================================
+#  Problem 2 - `mean, median`
+
+# implement methods "mean", "median" on Array of numbers
+def mean(ary)
+  # get sum
+  sum = ary.reduce(0.0) {|item, acc| acc + item}
+  
+  # divide by the numbers of element
+  return sum/ary.count
+end
+
+def median(ary)
+  n = ary.count
+  m = nil
+
+  #sort
+  sorted_ary = ary.sort
+
+  #if ary has even number of element return average of element count/2 and count/2 
+  if(n % 2 == 0)
+    sum = sorted_ary.at(n/2-1) + sorted_ary.at(n/2)
+    m = sum/2
+  else
+    #else return value of element count/2
+    m = sorted_ary.at(n/2)  
+  end
+
+  return m
+end
+
+# Your method should generate the following results:
+mean [1, 2, 3]    #=> 2.0
+mean [1, 1, 4]    #=> 2.0
+
+median [1, 2, 3]  #=> 2
+median [1, 1, 4]  #=> 1
+
+
+# ========================================================================================
+#  Problem 3 - `pluck`
+
+# implement method `pluck` on array of hashes
+def pluck(ary, key)
+  key_ary =[]
+  ary.map {|item| key_ary << item[key]}
+  return key_ary
+end
+
+# Your method should generate the following results:
+records = [
+  {name: "John",   instrument: "guitar"},
+  {name: "Paul",   instrument: "bass"  },
+  {name: "George", instrument: "guitar"},
+  {name: "Ringo",  instrument: "drums" }
+]
+pluck records, :name        #=> ["John", "Paul", "George", "Ringo"]
+pluck records, :instrument  #=> ["guitar", "bass", "guitar", "drums"]
+
+
+# ========================================================================================
+#  Problem 4 - monthly bank statement
+
+# given a CSV file with bank transactions for a single account (see assignment02-input.csv)
+# generate an HTML file with a monthly statement
+
+# assume starting balance is $0.00
+
+# the monthly statement should include the following sections:
+# - withdrawals
+# - deposits
+# - daily balance
+# - summary:
+#   - starting balance, total deposits, total withdrawals, ending balance
+
+def create_transaction(date, payee, amount, type)
+  {date: date, payee: payee, amount: amount, type: type}
+end
+
+#get records based on type
+def get_records(records, type)
+  r = records.select{ |r| r[:type] == type}
+  r.sort_by!{|r| r[:date].split('/').map{|d| d.to_i} }
+  return r
+end
+
+#get sum of all records
+def records_sum(records)
+  sum = 0.0;
+  records.each{|r| sum+= r[:amount].to_f}
+  return sum
+end
+
+def get_daily_balances(records)
+  daily_balance = []
+  balance = 0.0
+  #get unique date array
+  dates = records.map{|r| r[:date]}.uniq
+  #sort date 
+  dates.sort_by!{|d| d.split('/').map{|d| d.to_i} }
+
+  dates.each do |d|
```

Comment:
use map instead of iteration via each. in each iteration of map, you calculate the next value, and don't worry about appending it to an array, since map will do that for you.

the big benefit of using map over doing iteration, is that you can then chain operations together.